actionpack/lib/action_dispatch/journey/formatter.rb

```diff
@@ -125,33 +125,19 @@ def non_recursive(cache, options)
           routes
         end
 
-        module RegexCaseComparator
-          DEFAULT_INPUT = /[-_.a-zA-Z0-9]+\/[-_.a-zA-Z0-9]+/
-          DEFAULT_REGEX = /\A#{DEFAULT_INPUT}\Z/
-
-          def self.===(regex)
-            DEFAULT_INPUT == regex
-          end
-        end
-
         # Returns an array populated with missing keys if any are present.
         def missing_keys(route, parts)
           missing_keys = nil
-          tests = route.path.requirements
+          tests = route.path.requirements_for_missing_keys_check
           route.required_parts.each { |key|
             case tests[key]
             when nil
               unless parts[key]
                 missing_keys ||= []
                 missing_keys << key
               end
-            when RegexCaseComparator
```

Comment:
How `RegexCaseComparator` is not in use anymore? It was being used here. That module was the equivalent to `if test[key] == RegexCaseComparator::DEFAULT_INPUT`

Reply:
Sorry, my wording was not quite correct. `RegexCaseComparator` achieves the same thing what the memoization does (with less allocations):

```ruby
          DEFAULT_INPUT = /[-_.a-zA-Z0-9]+\/[-_.a-zA-Z0-9]+/
          DEFAULT_REGEX = /\A#{DEFAULT_INPUT}\Z/
```

vs

https://github.com/rails/rails/pull/38184/files#diff-67940fe9af75d2f6ec9596b2d13998c7R181

The only difference is that we memoize regexes per path. I scan the codebase and couldn't find a case when `RegexCaseComparator` is used. I couldn't find the regex referenced either:

`ag --literal 'RegexCaseComparator'`
`ag --literal '[-_.a-zA-Z0-9]+\/[-_.a-zA-Z0-9]+'`

I can add the module back with a few test cases so we have some coverage. What do you think?

Reply:
Looking the commit that introduced that module it was a way to make a similar optimization as this PR but to save memory. https://github.com/rails/rails/commit/097ec6fb7c7e1854cdd96113213b555ae7415953.

With your change we are going back to use that amount of memory if the assumption that most routes have a `route.path.requirements[key]` of `/[-_.a-zA-Z0-9]+\/[-_.a-zA-Z0-9]+/` is correct since instead of a single instance we would have N `DEFAULT_REGEX`'s instances.

Reply:
@schneems where did you find that most requirements are `/[-_.a-zA-Z0-9]+\/[-_.a-zA-Z0-9]+/`? Running the Rails test suite I found no reference for that regular expression.

Reply:
It seems it was from here https://github.com/codetriage/codetriage/blob/master/config/routes.rb#L59. Did we just optimized Rails to a single app? 😂

Reply:
thanks @rafaelfranca :bowing_man: